File: ios/Classes/Features/Request Tracking/ManualRequestTracker.swift
import ADEUMInstrumentation
import Flutter

extension SwiftAppDynamicsAgentPlugin {
  func getRequestTrackerWithUrl(result: @escaping FlutterResult, arguments: Any?) {
    let properties = arguments as! Dictionary<String, Any>
    let id = properties["id"] as! String
    
    guard let urlString = properties["url"] as? String, let url = URL(string: urlString) else {
      let error = FlutterError(code: "500", message: "Agent getRequestTrackerWithUrl() failed.", details: "Please provide a valid URL.")
      result(error)
      return
    }
    
    let requestTracker: ADEumHTTPRequestTracker = ADEumHTTPRequestTracker(url: url)
    requestTrackers[id] = requestTracker
    
    result(nil)
  }
  
  func setRequestTrackerErrorInfo(result: @escaping FlutterResult, arguments: Any?) {
    let properties = arguments as! Dictionary<String, Any>
    let id = properties["id"] as! String
    let errorDict = properties["errorDict"] as! Dictionary<String, String>
    
    guard let tracker = requestTrackers[id] else {
      let error = FlutterError(code: "500", message: "Agent setRequestTrackerErrorInfo() failed.", details: "Request tracker was not initialized or already reported.")
      result(error)
      return
    }
    
    tracker.error = NSError(domain: "Manual request tracker", code: 0, userInfo: errorDict)
    result(nil)
  }
  
  func setRequestTrackerStatusCode(result: @escaping FlutterResult, arguments: Any?) {
    let properties = arguments as! Dictionary<String, Any>
    let id = properties["id"] as! String
    
    guard let tracker = requestTrackers[id] else {
      let error = FlutterError(code: "500", message: "Agent setRequestTrackerStatusCode() failed.", details: "Request tracker was not initialized or already reported.")
      result(error)
      return
    }
    
    guard let statusCode = properties["statusCode"] as? NSNumber else {
      let error = FlutterError(code: "500", message: "Agent setRequestTrackerStatusCode() failed.", details: "Please provide a valid status code.")
      result(error)
      return
    }
    
    tracker.statusCode = statusCode
    result(nil)
  }
  
  func setRequestTrackerResponseHeaders(result: @escaping FlutterResult, arguments: Any?) {
    let properties = arguments as! Dictionary<String, Any>
    let id = properties["id"] as! String
    
    guard let tracker = requestTrackers[id] else {
      let error = FlutterError(code: "500", message: "Agent setRequestTrackerResponseHeaders() failed.", details: "Request tracker was not initialized or already reported.")
      result(error)
      return
    }
    
    guard let headers = properties["headers"] as? [String: [String]] else {
      let error = FlutterError(code: "500", message: "Agent setRequestTrackerResponseHeaders() failed.", details: "Please provide valid response headers.")
      result(error)
      return
    }
    
    let stringValues = headers.mapValues { value in
      return value.joined(separator: ", ")
    }
    
    tracker.allHeaderFields = stringValues
    result(nil)
  }
  
  func setRequestTrackerRequestHeaders(result: @escaping FlutterResult, arguments: Any?) {
    let properties = arguments as! Dictionary<String, Any>
    let id = properties["id"] as! String
    
    guard let tracker = requestTrackers[id] else {
      let error = FlutterError(code: "500", message: "Agent setRequestTrackerRequestHeaders() failed.", details: "Request tracker was not initialized or already reported.")
      result(error)
      return
    }
    
    guard let headers = properties["headers"] as? [String: [String]] else {
      let error = FlutterError(code: "500", message: "Agent setRequestTrackerRequestHeaders() failed.", details: "Please provide valid response headers.")
      result(error)
      return
    }
    
    let stringValues = headers.mapValues { value in
      return value.joined(separator: ", ")
    }
    
    tracker.allRequestHeaderFields = stringValues
    result(nil)
  }
  
  func requestTrackerReport(result: @escaping FlutterResult, arguments: Any?) {
    let properties = arguments as! Dictionary<String, Any>
    let id = properties["id"] as! String
    
    guard let tracker = requestTrackers[id] else {
      let error = FlutterError(code: "500", message: "Agent requestTrackerReport() failed.", details: "Request tracker was not initialized or already reported.")
      result(error)
      return
    }
    
    tracker.reportDone()
    requestTrackers[id] = nil
    
    result(nil)
  }
  
  func getServerCorrelationHeaders(result: @escaping FlutterResult, arguments: Any?) {
    let headers = ADEumServerCorrelationHeaders.generate()
    let listValues = Dictionary(uniqueKeysWithValues:
                                  headers.map { key, value in (key, [value]) })
    
    result(listValues)
  }
}

File: android/src/main/kotlin/com/appdynamics/appdynamics_agent/features/request_tracking/ManualRequestTracker.kt
/*
 * Copyright (c) 2022. AppDynamics LLC and its affiliates.
 * All rights reserved.
 *
 */

package com.appdynamics.appdynamics_agent.features.request_tracking

import androidx.annotation.NonNull
import com.appdynamics.appdynamics_agent.AppDynamicsAgentPlugin
import com.appdynamics.eumagent.runtime.Instrumentation
import com.appdynamics.eumagent.runtime.ServerCorrelationHeaders
import io.flutter.plugin.common.MethodChannel
import java.net.MalformedURLException
import java.net.URL

fun AppDynamicsAgentPlugin.getRequestTrackerWithUrl(
    @NonNull result: MethodChannel.Result,
    arguments: Any?
) {
    val properties = arguments as HashMap<*, *>
    val id = properties["id"] as String

    val urlString = properties["url"] as? String ?: run {
        result.error(
            "500",
            "Agent getRequestTrackerWithUrl() failed.",
            "Please insert valid URL string."
        )
        return
    }

    val url = URL(urlString)
    requestTrackers[id] = Instrumentation.beginHttpRequest(url)
    result.success(null)
}

fun AppDynamicsAgentPlugin.setRequestTrackerErrorInfo(
    @NonNull result: MethodChannel.Result,
    arguments: Any?
) {
    val properties = arguments as HashMap<*, *>
    val id = properties["id"] as String

    val error = properties["errorDict"] as? HashMap<String, String> ?: run {
        result.error(
            "500",
            "Agent setRequestTrackerErrorInfo() failed.",
            "Please insert a valid error message."
        )
        return
    }

    val tracker = requestTrackers[id] ?: run {
        result.error(
            "500",
            "Agent setRequestTrackerErrorInfo() failed.",
            "Request tracker was not initialized or already reported."
        )
        return
    }

    tracker.withError(error["message"])
    result.success(null)
}

fun AppDynamicsAgentPlugin.setRequestTrackerStatusCode(
    @NonNull result: MethodChannel.Result,
    arguments: Any?
) {
    val properties = arguments as HashMap<*, *>
    val id = properties["id"] as String

    val tracker = requestTrackers[id] ?: run {
        result.error(
            "500",
            "Agent setRequestTrackerStatusCode() failed.",
            "Request tracker was not initialized or already reported."
        )
        return
    }

    val statusCode = properties["statusCode"] as? Int ?: run {
        result.error(
            "500",
            "Agent setRequestTrackerStatusCode() failed.",
            "Status code must be an integer."
        )
        return
    }

    tracker.withResponseCode(statusCode)
    result.success(null)
}

fun AppDynamicsAgentPlugin.setRequestTrackerResponseHeaders(
    @NonNull result: MethodChannel.Result,
    arguments: Any?
) {
    val properties = arguments as HashMap<*, *>
    val id = properties["id"] as String

    val tracker = requestTrackers[id] ?: run {
        result.error(
            "500",
            "Agent setRequestTrackerResponseHeaders() failed.",
            "Request tracker was not initialized or already reported."
        )
        return
    }

    val headers = properties["headers"] as? Map<String, List<String>> ?: run {
        result.error(
            "500",
            "Agent setRequestTrackerResponseHeaders() failed.",
            "Headers are not of type Map<String, String>."
        )
        return
    }

    tracker.withResponseHeaderFields(headers)
    result.success(null)
}

fun AppDynamicsAgentPlugin.setRequestTrackerRequestHeaders(
    @NonNull result: MethodChannel.Result,
    arguments: Any?
) {
    val properties = arguments as HashMap<*, *>
    val id = properties["id"] as String

    val tracker = requestTrackers[id] ?: run {
        result.error(
            "500",
            "Agent setRequestTrackerRequestHeaders() failed.",
            "Request tracker was not initialized or already reported."
        )
        return
    }

    val headers = properties["headers"] as? Map<String, List<String>> ?: run {
        result.error(
            "500",
            "Agent setRequestTrackerRequestHeaders() failed.",
            "Headers are not of type Map<String, String>."
        )
        return
    }

    tracker.withRequestHeaderFields(headers)
    result.success(null)
}

fun AppDynamicsAgentPlugin.requestTrackerReport(
    @NonNull result: MethodChannel.Result,
    arguments: Any?
) {
    val properties = arguments as HashMap<*, *>
    val id = properties["id"] as String

    val tracker = requestTrackers[id] ?: run {
        result.error(
            "500",
            "Agent requestTrackerReport() failed.",
            "Request tracker was not initialized or already reported."
        )
        return
    }

    tracker.reportDone()
    result.success(null)
}

fun AppDynamicsAgentPlugin.getServerCorrelationHeaders(
    @NonNull result: MethodChannel.Result,
    arguments: Any?
) {
    val headers = ServerCorrelationHeaders.generate()
    result.success(headers)
}